File: sample/sample.cpp
#include <iostream>
#include <string>
#include <simstring/simstring.h>

void retrieve(
    simstring::reader& dbr,
    const std::string& query,
    int measure,
    double threshold
    )
{
    // Retrieve similar strings into a string vector.
    std::vector<std::string> xstrs;
    dbr.retrieve(query, measure, threshold, std::back_inserter(xstrs));

    // Output the retrieved strings separated by ", ".
    for (int i = 0;i < (int)xstrs.size();++i) {
        std::cout << (i != 0 ? ", " : "") << xstrs[i];
    }
    std::cout << std::endl;
}

int main(int argc, char *argv[])
{
    // Create a SimString database with two person names.
    simstring::ngram_generator gen(3, false);
    simstring::writer_base<std::string> dbw(gen, "sample.db");

    dbw.insert("Barack Hussein Obama II");
    dbw.insert("James Gordon Brown");
    dbw.close();

    // Open the database for reading.
    simstring::reader dbr;
    
    dbr.open("sample.db");
    retrieve(dbr, "Barack Obama", simstring::cosine, 0.6);
    retrieve(dbr, "Gordon Brown", simstring::cosine, 0.6);
    retrieve(dbr, "Obama", simstring::cosine, 0.6);
    retrieve(dbr, "Obama", simstring::overlap, 1.0);

    return 0;
}

File: sample.py
#!/usr/bin/env python

import simstring

# Create a SimString database with two person names.
db = simstring.writer('sample.db')
db.insert('Barack Hussein Obama II')
db.insert('James Gordon Brown')
db.close()


# Open the database for reading.
db = simstring.reader('sample.db')

# Use cosine similarity and threshold 0.6.
db.measure = simstring.cosine
db.threshold = 0.6
print(db.retrieve('Barack Obama'))      # OK.
print(db.retrieve('Gordon Brown'))      # OK.
print(db.retrieve('Obama'))             # Too dissimilar!

# Use overlap coefficient and threshold 1.0.
db.measure = simstring.overlap
db.threshold = 1.
print(db.retrieve('Obama'))             # OK.
